Notify Receptionist subscribers when node added and removed, #28792

patriknw opened the pull request:
* If the change from ddata arrives before the join information
* One difficulty is that that the removal tick may trigger removal
  for entries that are in the ddata state but not in the membership yet.
  Added a delay to solve that.

References #28792

patriknw committed:
Notify Receptionist subscribers when node added and removed, #28792

* If the change from ddata arrives before the join information
* One difficulty is that that the removal tick may trigger removal
  for entries that are in the ddata state but not in the membership yet.
  Added a delay to solve that.

patriknw commented on akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala:
```diff
@@ -528,14 +561,33 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
-              nodesRemoved(notInCluster)
+              // FIXME config of delay
+              import scala.concurrent.duration._
+              ctx.scheduleOnce(5.seconds, ctx.self, CleanupRemoved(notInCluster))
```
This is something I'm uncertain about. Can we solve this in a better way?
The problem is that we might have entries from ddata but not the membership information yet, so it looks like stale entries that should be removed.

johanandren replied:
I don't quite get this scenario, the periodic tick would notice nodes in cluster that were removed, but here instead we defer it wit ha delay so that a ddata remove can arrive inbetween? I don't see the problem with the service being removed "twice".

patriknw replied:
The problem is when the tick happens before we have seen the node as joining, but we have already got the ddata update. Then it looks like we have an entry in the ddata registry that is supposed to be removed, because the node isn't there.

patriknw commented on akka-cluster-typed/src/test/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionistSpec.scala:
```diff
@@ -111,6 +113,33 @@ class ClusterReceptionistSpec extends AnyWordSpec with Matchers with LogCapturin
       }
     }
 
+    "handle registrations before joining" in {
```
this was not a problem, just added to be sure

patriknw commented on akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala:
```diff
@@ -304,14 +305,21 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
 
       def reachabilityChanged(keysForNode: Set[AbstractServiceKey], newState: State): Unit = {
-        keysForNode.foreach { changedKey =>
+        notifySubscribers(keysForNode, servicesWereAddedOrRemoved = false, newState)
```
I have to check what servicesWereAddedOrRemoved is used for, but this is how it was before. Reachable/Unreachable is not changing allServiceInstances.

patriknw replied:
The theoretically possible scenario would be:
* ddata update seen first, entry added to the registry but corresponding node not seen yet
* UnreachableMember arrives before MemberJoined, the unreachable can be for any node not just for the joining

Hm, this is difficult to keep in sync. I was thinking if we could keep track of previously published Listing and compare, but it's for each key and also subscribers can be added after the previously published Listing.

This is like an optimization hint and we can error on the side of servicesWereAddedOrRemoved=true even though it was the same allServiceInstances as the previous. Maybe we can use that in some way. Have to think some more about it.

patriknw replied:
Second thought, that scenario isn't a problem. It will publish a Listing with servicesWereAddedOrRemoved=true when the MemberJoined arrives. I don't think this is a problem.

patriknw commented on akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala:
```diff
@@ -476,15 +478,44 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
+            Behaviors.same
+          } else {
+            val newState = state.copy(registry = state.registry.addNode(uniqueAddress))
+            val keysForNode = newState.registry.keysFor(uniqueAddress)
```
> we could already have services registered for the node (which we kind of do not know of)

That is right. The ddata update arrives before the membership gossip.

> couldn't a better way around that being handling it when we get the ddata update and notice the node is unknown to us

I'm not sure what we would do then? Stash the ddata update somewhere until member is known. Adding complexity.

johanandren replied:
I was thinking we can just add the node upon seeing the ddata update if node is unknown, but perhaps that clashes with the new-node-incarnation-handling.

patriknw commented on akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala:
```diff
@@ -538,7 +573,8 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
               nodesRemoved(notInCluster)
```
No delay here anymore.

patriknw commented on akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala:
```diff
@@ -522,14 +558,13 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
-          // ok to update from several nodes but more efficient to try to do it from one node
-          if (isLeader) {
+          // ok to update from several nodes but more efficient to not do the same from all
+          if (ThreadLocalRandom.current().nextInt(state.registry.nodes.size) == 0) {
```
Can't rely on leader here, because a completely new node may be the leader (lowest address) and then it doesn't know about `removedNodes`. Spreading the responsibility of the cleanup over all nodes.

patriknw replied:
no, this isn't completely safe. It's possible that all of the remaining nodes that know about the removedNodes are also removed. That may result in entries that are not removed by this RemoveTick process.

I have to take this for another round.

patriknw commented on akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala:
```diff
@@ -476,15 +475,50 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
+            // RemoteTick should only remove entries for removed nodes. It's possible
+            // that an entry is added before the node has been visible in the membership.
+            // DData replication may happen before membership gossip.
+            val cleanupRemovedNodeAfter = settings.pruningInterval * 100
```
this comment is relevant for the RemoveTick problem

patriknw reviewed (commented):
@johanandren Updated with new approach for the RemoveTick problem.

patriknw committed:
3rd solution for the RemoveTick problem

* it's possible that the Entry arrives before MemberJoined, but
  such entries should not be removed by the RemoveTick
* by adding a timestamp of when the entry is created we can
  avoid removing such early entries and the MemberJoined should be
  visible before that duration has elapsed

patriknw commented:
Updated with 3rd solution for the RemoveTick problem.

johanandren reviewed (approved):
LGTM 👍

renatocaval commented on akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala:
```diff
@@ -476,25 +481,54 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
             // watch-Terminated or will be removed by other nodes. This point is anyway too late.
             Behaviors.stopped
-          } else {
+          } else if (state.registry.nodes(uniqueAddress)) {
```
```suggestion
          } else if (state.registry.nodes.contains(uniqueAddress)) {
```

octonato reviewed (approved):
LGTM

Left two minor (nitpicking) suggestions. Not required, fine by me if merged as such.

patriknw commented on akka-cluster-typed/src/main/scala/akka/cluster/typed/internal/receptionist/ClusterReceptionist.scala:
```diff
@@ -476,25 +481,54 @@ private[typed] object ClusterReceptionist extends ReceptionistBehaviorProvider {
 
         case NodeAdded(uniqueAddress) =>
-          behavior(setup, state.copy(registry = state.registry.addNode(uniqueAddress)))
+          if (state.registry.nodes(uniqueAddress)) {
```
I don't mind making that more explicit

patriknw committed:
Apply suggestions from code review

Co-Authored-By: Renato Cavalcanti <renato@cavalcanti.be>
